### PR TITLE
Albedo from file

### PR DESCRIPTION
**Pull request by kmdeck**
## For reviewers
This is a big PR - the major files changed are src/Bucket/Bucket.jl, src/Regridder.jl, and src/SharedUtilities/models.jl.

## Purpose 
This PR adds in the capability to read in a NetCDF file of surface albedo from a file, store in the the auxiliary state. Currently this only supports fields that do not vary in time; future PRs will address updating the albedo at each step/ at each X number of steps, and will ensure that the correct entry is being read in and used for the first timestep, based on t0.

This has been tested with the coupler in both MPI and non-MPI runs.

## Content
- Added Regridder utils from ClimaCoupler. Longer term, perhaps we should have one copy of these tools, living in a separate repo (ClimaCoreTempestRemap?), to avoid duplication.
- Created artifacts.jl file in `src/Bucket/artifacts` which can stored the needed artifacts for the global bucket simulation.
- Created a new struct "MapInfo", which stores needed information for reading in netcdf files, regridding and making a field of them. Any parameter which is read from a file will longer term be passed in as type ::MapInfo, rather than e.g. `FT` or a `Function`.
- Created an albedo model type which uses `MapInfo` rather than a prescribed `f::Function` for albedo = f(lat,lon). The two options are now `BulkAlbedoFunction` (where soil albedo is a function) and `BulkAlbedoMap`, where soil albedo is of type `MapInfo`.
- Added soil albedo to auxiliary variables. For both albedo models, we compute the albedo once, up front (see next item), and store in the auxiliary state. If the albedo needs to be updated in time, the `update_aux` function is ready to be used for that.
- Created a `set_initial_aux_state!` function which inits the aux variables (based on initial conditions for the prognostic state) as well as inits any parameters read from a dataset that are also stored in aux
- Added partial testing of new capabilities. The Regridder utils are already tested in ClimaCoupler. Longer term, we should import these functions rather than redefine (and retest) here.
- Updated all bucket tests and tutorials to use the new albedo types.

Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

File: src/Regridder.jl
Comment (LenkaNovak) [inline src/Regridder.jl:1]: Given that a lot of this is copied from ClimaCoupler, it would be preferable to move this to ClimaCore to avoid duplication. @simonbyrne, what do you think?

File: docs/tutorials/Bucket/bucket_tutorial.jl
Comment (braghiere) [inline docs/tutorials/Bucket/bucket_tutorial.jl:161]: α_soil should be α_surface since this is a combination of soil+vegetation.

**Review by braghiere — COMMENTED** [review]
Just naming a variable differently to be precise (surface instead of soil), and clarifying the usage of snow vs. soil.

File: src/Bucket/Bucket.jl
Comment (juliasloan25) [inline src/Bucket/Bucket.jl:262]: This seems like it could maybe be done using multiple dispatch instead, though I'm not sure how we could access `parameters.albedo` in the function signature itself.

File: src/Bucket/Bucket.jl
Comment (juliasloan25) [inline src/Bucket/Bucket.jl:94]: Change to `\alpha_surface` per Renato's comment about the file albedo data.

File: src/Bucket/Bucket.jl
Comment (juliasloan25) [inline src/Bucket/Bucket.jl:386]: could unpack these with `(; ...)` syntax to condense this code

File: src/Regridder.jl
Comment (juliasloan25) [inline src/Regridder.jl:304]: this function is a duplicate of `read_from_hdf5`. `read_from_hdf5` just has a name update to be consistent with `write_to_hdf5`, and adds a default value for `comms_ctx`. It's the more updated version that will be used by the coupler going forward.

File: src/Regridder.jl
Comment (juliasloan25) [inline src/Regridder.jl:350]: change to `read_from_hdf5` (see above comment)

File: test/Bucket/albedo_map_bucket.jl
Comment (juliasloan25) [inline test/Bucket/albedo_map_bucket.jl:60]: could use `for bucket_domain in bucket_domains:` to improve readability

**Review by juliasloan25 — APPROVED** [review]

File: Project.toml
Comment (Yujie-W) [inline Project.toml:11]: I do not see corresponding items in the `[compat]` session

**Review by Yujie-W — CHANGES_REQUESTED** [review]
1. Add corresponding `[compat]` items in the `Project.toml` file. Otherwise, the registration will be redflagged in the future.
2. I am curious why not using Artifacts.jl and lazy artifacts file? Then you can simply use `artifact"ART_NAME"` in the code? [No changes required]
3. I see there are some lines such as FT(1.0) and FT(0.0). If 1 and 0 are directly used along with FT, such as `1 - ft_number`, you do not need to use `FT(1) - ft_number`. Does this result in any difference in speed? [No changes required]

File: src/Bucket/Bucket.jl
Comment (charleskawczynski) [reply src/Bucket/Bucket.jl:262]: We can use UnionSplitting
```suggestion
    if parameters.albedo isa BulkAlbedoMap
```
which is equivalent to multiple dispatch for small (i.e., a few) unions, ref: https://julialang.org/blog/2018/08/union-splitting/

Comment (kmdeck) [reply src/Bucket/Bucket.jl:262]: nice!

File: src/Regridder.jl
Comment (kmdeck) [reply src/Regridder.jl:304]: good catch!

**Comment by kmdeck** [comment]
> 1. Add corresponding `[compat]` items in the `Project.toml` file. Otherwise, the registration will be redflagged in the future.
> 
>     3. I see there are some lines such as FT(1.0) and FT(0.0). If 1 and 0 are directly used along with FT, such as `1 - ft_number`, you do not need to use `FT(1) - ft_number`. Does this result in any difference in speed? [No changes required]

good catch Yujie! I fixed the toml and at least some of the unnecessary FT() calls. will keep an eye out for those.

**Comment by kmdeck** [comment]
> 2\. I am curious why not using Artifacts.jl and lazy artifacts file? Then you can simply use `artifact"ART_NAME"` in the code? [No changes required]

With respect to this - Im not sure. ArtifactWrappers uses Artifacts.jl under the hood, with an option for a lazy download (https://clima.github.io/ArtifactWrappers.jl/dev/). I'm not sure we need the extra features the wrapper provides, but I think it might amount to the same core capabilities in the end. What do you think?

File: test/Bucket/albedo_map_bucket.jl
Comment (juliasloan25) [inline test/Bucket/albedo_map_bucket.jl:33]: It might be nice to have an additional `MapInfo` constructor which has default values for all the arguments, so a basic map can be created without having to know about the internals. The coupler has to create an albedo `MapInfo` struct and I think it would be better if it didn't have to use e.g. `cesm2_land_albedo_dataset_path()`. This could be a separate small PR, or maybe part of this one.

Comment (kmdeck) [reply test/Bucket/albedo_map_bucket.jl:33]: ok! I think the default constructor would need to call cesm2_land_albedo_dataset_path() - to get the local path of the file - but that's fine! Ill add this in today.

Comment (kmdeck) [reply test/Bucket/albedo_map_bucket.jl:31]: ok! this is fixed. you can now call it with albedo_model = BulkAlbedoMap{FT}(regrid_dirpath). It also has a default value for alpha snow, and comms context, so you'll have to pass in the comms context from the coupler.
https://github.com/CliMA/ClimaLSM.jl/blob/6d4e3b11a7b407011df0d76570f5c10865f57f98/test/Bucket/albedo_map_bucket.jl#L35
https://github.com/CliMA/ClimaLSM.jl/blob/6d4e3b11a7b407011df0d76570f5c10865f57f98/src/Bucket/Bucket.jl#L100

**Comment by kmdeck** [comment]
@Yujie-W I think you need to approve the review before I can merge (given that you requested changes). I left some comments back. Otherwise I think I can dismiss the review. let me know. thank you!

File: src/Regridder.jl
Comment (kmdeck) [reply src/Regridder.jl:1]: @simonbyrne @juliasloan25 perhaps we can merge this PR, and then a later PR would move the duplicate code from the coupler and climalsm into climacore?

**Review by Yujie-W — APPROVED** [review]

**Pull request merged**
